docs/source/components.rst

```diff
@@ -0,0 +1,332 @@
+per-request basis, e.g. "fetch the existing data from source A if user is admin,
+otherwise fetch it from source B".
+
+How it is currently done
+------------------------
+
+Form field defaults can be specified with the ``initial`` argument to
+``Field()``:
+
+.. code-block:: python
+
+    title = CharField(
+        max_length=200,
+        initial='Enter your awesome title here!'
+    )
+
+
+It is also possible to pass a dictionary of initial data to ``Form.__init__()``:
+
+.. code-block:: python
+
+    my_initial_data = {'title': 'Enter your awesome title here!'}
+    form = BlogPostForm(initial=my_initial_data)
+
+
+How it could be done
+--------------------
+
+See "How it could be done" on the section "New Data".
+
+
+New Data
+========
+
+All data that is provided by external sources (e.g. an HTML form submitted by a
+user or POSTed JSON data) is called "new data". For the moment we'll call them
+"Input Serializers". An Input Serializer must follow a documented API that
+allows data of all kinds and shapes to be converted to a universal data
+structure. That data structure then serves a data source for other components.
+
+How it is currently done
+------------------------
+
+In Django Forms user data is passed to a form instance via the ``data`` and
+``files`` argument:
+
+.. code-block:: python
+
+    form = BlogPostForm(data=request.POST, files=request.FILES)
+
+
+The conversion to a "universal" data structure then happens in
+``Widget.value_from_datadict()``. This method handles all potential oddities of
+an HTML form, like multiple input fields with the same name or different input
+fields belonging to one logical field.
+
+DRF uses the ``data`` argument to a Serializer:
+
+.. code-block:: python
+
+    serializer = BlogPostSerializer(data=json_data)
+
+
+How it could be done
+--------------------
+
+At first the universal data structure has to be defined. It should not be
+opinionated about how the input data looked like and ideally should be usable
+not only with new data but with existing data as well.
+
+Then an API must be established that lets you modify the loading of existing or
+new data easily.
+
+
+Validation
+==========
+
+The validation step comprises validating field values (e.g. field ``name`` must
+have between 5 and 10 alphanumerical characters or field ``publish_date`` must
+be in the future) and cross field validation (e.g. if field A is set field B
+must be unset and vice versa, or field ``first_number`` and field
+``second_number`` added up must be smaller than 100).
+
+This component works only with the universal data structure mentioned before and
+should not fetch additional data itself. It should however be possible for the
+validation component to be influenced by the environment (a user's permission,
+current time, etc.) and change its logic based on that.
+
+Additionally a validator should also be allowed to change values in a structural
```

Comment:
Let's call this "cleaning", in fact I'm possibly feeling the whole step should be called cleaning and non-changing things are validators. This is still a bit unclear though.

I'd really like to have the idea that "cross-field" validation/cleaning is not a second class citizen.